### PR TITLE
Support compiling wast to wasm in testeth

### PR DESCRIPTION
**Pull request by axic**
Depends on #4779.

### CODE REVIEW

**Comment by axic** [comment]
@chfast this need binaryen support in cpp-eth :)

cc @jwasinger @hugo-dc @holgerd77

@winsvega this should not be merged soon, it is only used by ewasm testing, can be merged once all of that is stable

**Comment by axic** [comment]
Should follow the instructions from here: https://gitter.im/ethereum/tests?at=5a577e71d0514c785b1a2384

File: test/tools/libtesteth/TestHelper.cpp [hunk 1]
```diff
@@ -382,6 +382,35 @@ string compileLLL(string const& _code)
...
+  wasm::WasmBinaryWriter writer(&wasm, buffer, debug);
+  writer.write();
+
+  ostringstream output;
+  buffer.writeTo(output);
+
+  return bytes(output.begin(), output.end());
+}
+
 void checkHexHasEvenLength(string const& _str)
```
Comment (winsvega) [inline test/tools/libtesteth/TestHelper.cpp]: do not change importCode function. 
the changes must be done to replaceLLL function 
but after this PR got merged please https://github.com/ethereum/cpp-ethereum/pull/4735/files

just rename replaceLLL function to compileSrcCode or something
and parse wasm and LLL from there

**Comment by winsvega** [comment]
is it because of a compiler version I use. or something is wrong with the dependency source files?

**Comment by axic** [comment]
Will rebase this on top off #4779 and I'll need to only change `replaceBytecode` to support wasm.

**Comment by axic** [comment]
@chfast this also needs to work when hera is turned off

File: test/tools/libtesteth/TestHelper.cpp (hunk 1 above)
Comment (axic) [reply test/tools/libtesteth/TestHelper.cpp]: Has been updated now.

**Comment by axic** [comment]
@winsvega do you have more context for this error?

```
/home/wins/Ethereum/cpp-ethereum/build/deps/src/binaryen/src/literal.h:158: error: specialization of ‘template<class _Tp> struct std::hash’ in different namespace [-fpermissive]
 template<> struct hash<wasm::Literal> {
                   ^
```

Like the preceding/following 40 lines.

**Comment by winsvega** [comment]
I think its the first one. because of this the following errors are due to the missing defenition.

**Comment by axic** [comment]
@winsvega without more context it is really hard to figure this out. Can you upload logs somewhere?

**Comment by winsvega** [comment]
selection_001

**Comment by axic** [comment]
@winsvega fixed all the build issues - please retry

**Comment by winsvega** [comment]
looks like it works now, but
```
/home/wins/Ethereum/cpp-ethereum/test/tools/libtesteth/TestHelper.cpp:268: error: undefined reference to 'dev::test::wast2wasm(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool)'
collect2: error: ld returned 1 exit status
test/CMakeFiles/testeth.dir/build.make:1798: recipe for target 'test/testeth' failed
make[2]: *** [test/testeth] Error 1
CMakeFiles/Makefile2:1269: recipe for target 'test/CMakeFiles/testeth.dir/all' failed
make[1]: *** [test/CMakeFiles/testeth.dir/all] Error 2
Makefile:140: recipe for target 'all' failed
make: *** [all] Error 2
```

what was the issue with binaryen?

**Comment by axic** [comment]
@winsvega you should rerun cmake to regenerate dependency lists.

**Comment by winsvega** [comment]
that didn't help

**Comment by axic** [comment]
@winsvega fixed! Please try with the latest version :)

**Comment by winsvega** [comment]
worked. 
how did you fixed the -fpermissive issue?

**Comment by axic** [comment]
Moved it into its own file so it won't collide with the include mess we have.

**Comment by winsvega** [comment]
so its like we have our own version of this dependency now?

**Comment by axic** [comment]
There is no new build dependency (binaryen is included via hera). What is the mac build issue?

**Comment by winsvega** [comment]
```
-- Build files have been written to: /Users/distiller/build/deps/src/binaryen-build
ninja: error: mkdir(/lib): Permission denied
ninja: build stopped: .
Exited with code 1
```

**Comment by winsvega** [comment]
could you comment what was the issue with binaryen ?
https://github.com/WebAssembly/binaryen/issues/1362#issuecomment-358098858

**Comment by axic** [comment]
I have no idea. Moved it to a file on its own where the other 40+ testeth headers are not included. Anything down the dependency chain could have caused it.

**Comment by winsvega** [comment]
windows build fails
```
wast2wasm.cpp
c:\projects\cpp-ethereum\build\deps\src\binaryen\src\parsing.h(175): error C2220: warning treated as error - no 'object' file generated (compiling source file C:\projects\cpp-ethereum\test\tools\libtesteth\wast2wasm.cpp) [C:\projects\cpp-ethereum\build\test\testeth.vcxproj]
c:\projects\cpp-ethereum\build\deps\src\binaryen\src\parsing.h(175): warning C4146: unary minus operator applied to unsigned type, result still unsigned (compiling source file C:\projects\cpp-ethereum\test\tools\libtesteth\wast2wasm.cpp) [C:\projects\cpp-ethereum\build\test\testeth.vcxproj]
```

**Comment by axic** [comment]
@chfast do not forget to apply the same MSVC change to hera (and evm2wasm).

**Comment by winsvega** [comment]
also need to check that wasm test fillers are compiled correctly

**Comment by axic** [comment]
@winsvega can you test it? Would be nice merging this.

**Comment by winsvega** [comment]
yaml compiler works but the code starts with 0x00 byte. which means stop execution.

**Comment by axic** [comment]
> yaml compiler works but the code starts with 0x00 byte. which means stop execution.

It starts with a wasm magic (`\x00asm`).

**Comment by winsvega** [comment]
yaml tests from test repo fails on this bytecode. looks like it does not recognize wasm bytecode.

**Comment by axic** [comment]
> yaml tests from test repo fails on this. looks like it does not recognize wasm bytecode.

I am not sure what you mean. What test case are you compiling with this? The ones from @jwasinger ? In any case, this patch should have zero effect on the current EVM test suite (is this the case?) and it will only do conversion on eWASM test cases and the output of those cannot be interpreted by EVM (hence we'll not merge the eWASM test suite into the main tests repo anytime soon).

**Comment by winsvega** [comment]
`the output of those cannot be interpreted by EVM`
ok.

File: test/tools/libtesteth/wast2wasm.cpp [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+/* from https://github.com/ewasm/evm2wasm */
+string wast2wasm(string const& input, bool debug)
+{
+    wasm::Module module;
+
+    try {
+        // FIXME: binaryen 1.37.28 actually modifies the input...
+        //        as a workaround make a copy here
+        string tmp = input;
```
[the remark this answers is not in the document]
Comment (chfast) [reply test/tools/libtesteth/wast2wasm.cpp]: I confirm `tmp` is a copy of `input`.

File: test/tools/libtesteth/wast2wasm.cpp [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        wasm::SExpressionParser parser(const_cast<char*>(tmp.c_str()));
```
Comment (chfast) [inline test/tools/libtesteth/wast2wasm.cpp]: You don't need the cast now.

Comment (axic) [reply test/tools/libtesteth/wast2wasm.cpp]: Updated.

Comment (axic) [reply test/tools/libtesteth/wast2wasm.cpp]: A cast is still needed as `c_str()` returns `const char*` and `SExpressionParser` wants `char*`. Do you suggest any other cast / solution?

**Comment by axic** [comment]
I was trigger happy - merging this now.

**Comment by chfast** [comment]
You probably could have used `.data()` instead of `.c_str()`.

**Comment by axic** [comment]
For `.data()` the spec says it is still `const char*`, but it is not guaranteed to be zero terminated. That binaryen class needs zero terminated c-strings.

**Comment by axic** [comment]
@chfast can you please push the MSVC workaround to evm2wasm and hera?

**Pull request merged**